src/control/JacobianController.h

```diff
@@ -0,0 +1,139 @@
+#pragma once
+
+#include "../filters/StateSpaceUtil.h"
+#include "../navtypes.h"
+#include "../world_interface/data.h"
+#include "TrapezoidalVelocityProfile.h"
+
+#include <array>
+#include <functional>
+#include <optional>
+
+#include <Eigen/Core>
+
+namespace control {
+
+/**
+ * @brief This class controls the position of a multidimensional mechanism
+ * with a trapezoidal velocity profile using the jacobian of the kinematics.
+ *
+ * @tparam outputDim The dimension of the output vector of the kinematic function.
+ * @tparam inputDim The dimension of the input vector of the kinematic function.
+ * @see https://en.wikipedia.org/wiki/Jacobian_matrix_and_determinant
+ */
+template <int outputDim, int inputDim> class JacobianController {
+public:
+	/**
+	 * @brief Construct a new controller.
+	 *
+	 * @param kinematicsFunc The kinematics of the mechanism being controlled. This is
+	 * a function that takes in some input vector (ex: joint angles of an arm) and outputs
+	 * another vector (ex: the 3d pose of the hand)
+	 * @param jacobianFunc The jacobian of the kinematic function.
+	 * When executing a command, getCommand() should be called at this rate.
+	 * @param maxVels The maximum velocity in the output space of the kinematic function.
+	 * @param maxAccels The maximum acceleration in the output space of the kinematic function.
+	 */
+	JacobianController(
+		const std::function<navtypes::Vectord<outputDim>(const navtypes::Vectord<inputDim>&)>&
+			kinematicsFunc,
+		const std::function<navtypes::Matrixd<outputDim, inputDim>(
+			const navtypes::Vectord<inputDim>&)>& jacobianFunc,
+		double maxVel, double maxAccel)
+		: kinematicsFunc(kinematicsFunc),
+		  jacobianFunc(jacobianFunc
+						   ? jacobianFunc
+						   : std::bind(filters::statespace::numericalJacobian, kinematicsFunc,
+									   std::placeholders::_1, outputDim)),
+		  velocityProfile(maxVel, maxAccel) {
+		assert(this->kinematicsFunc);
+		assert(this->jacobianFunc);
+	}
+
+	/**
+	 * @brief Get the current command.
+	 *
+	 * @param currTime The current timestamp.
+	 * @param currPos The current position, in the input space.
+	 * @return navtypes::Vectord<inputDim> The target position, in the input space of the
+	 * kinematic function. Returns the current position if no target is set.
+	 */
+	navtypes::Vectord<inputDim> getCommand(robot::types::datatime_t currTime,
+										   const navtypes::Vectord<inputDim>& currPos) const {
```

Comment:
Again, indentation seems to be a little too large here (sorry I'm a stickler for this haha)